### PR TITLE
Changes to item upgrade system

### PR DESCRIPTION
**Pull request by GingerAvalanche**
Retry of #732, this time without being a 2 year old branch.

Implements Issue #93 changes
Implements Issue #289 changes

Allows any item type to be upgraded, and allows for upgrade slot handling on a per-state basis.

### CODE REVIEW

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc [hunk 1]
```diff
@@ -4651,6 +4651,7 @@ static function UpgradeItems(XComGameState NewGameState, name CreatorTemplateNam
 			if (!XComHQ.HasItem(UpgradeItemTemplate))
 			{
 				UpgradedItemState = UpgradeItemTemplate.CreateInstanceFromTemplate(NewGameState);
+				class'X2EventManager'.static.GetEventManager().TriggerEvent('ItemUpgraded', UpgradedItemState, none, NewGameState); // Issue #289 - handle upgrades for infinite items
```
Comment (pledbrook) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc]: Please use
```
`XEVENTMGR.TriggerEvent(...)
```
in place of
```
class'X2EventManager'.static.GetEventManager().TriggerEvent(...)
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc [hunk 2]
```diff
@@ -4668,6 +4669,7 @@ static function UpgradeItems(XComGameState NewGameState, name CreatorTemplateNam
 					// Otherwise match the base items quantity
 					UpgradedItemState = UpgradeItemTemplate.CreateInstanceFromTemplate(NewGameState);
 					UpgradedItemState.Quantity = BaseItemState.Quantity;
+					class'X2EventManager'.static.GetEventManager().TriggerEvent('ItemUpgraded', UpgradedItemState, BaseItemState, NewGameState); // Issue #289 - handle upgrades for utility items like grenades and medkits
```
Comment (pledbrook) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc]: Same goes here and anywhere else in the pull request.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc (hunk 1 above)
Comment (pledbrook) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc]: In addition, these events will need [highlander docs](https://x2communitycore.github.io/X2WOTCCommunityHighlander/#documentation-for-the-documentation-tool) associated with them. There are examples in various pull requests and in the current `master` branch of the highlander (I recommend rebasing this branch against it).

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc [hunk 3]
```diff
@@ -2363,6 +2363,27 @@ function bool IsMissionObjectiveItem()
 	return false;
 }
 
+// Start Issue #93
+// This function fires an event that mods can listen for which will allow them to modify upgrade slots on a per-gamestate basis
```
Comment (pledbrook) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc]: HL-Docs needed here too.

Comment (GingerAvalanche) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc]: Done. Also wrote the docs for #260 which needed them.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc (hunk 2 above)
Comment (GingerAvalanche) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc]: Done.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc (hunk 1 above)
Comment (GingerAvalanche) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc]: Done. https://github.com/X2CommunityCore/X2WOTCCommunityHighlander/pull/734/commits/d45d155bc6327454d846cfdb9503f83089a1d4be

**Comment by GingerAvalanche** [comment]
First time I've managed to rebase properly! \o/ All requested changes should be resolved, but I can't figure out how to get the pr to consider them resolved in the final summary.

**Comment by GingerAvalanche** [comment]
Alright, after about a hundred force-pushes because I didn't realize what files had what changes in any given commit snapshot, this PR now looks clean and should be easy to tell what it changes.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc [hunk 4]
```diff
@@ -4651,6 +4651,17 @@ static function UpgradeItems(XComGameState NewGameState, name CreatorTemplateNam
...
+				/// HL-Docs: feature:ItemUpgraded; issue:289; tags:strategy,events
+				/// This is an event that allows mods to perform non-standard handling to item states
+				/// when they are upgraded to a new tier. It is fired up to four times during upgrading:
+				/// * When upgrading the infinite item, if any.
+				/// * When upgrading utility items, like grenades and medkits.
+				/// * When upgrading unequipped items that have attachments
+				/// * When upgrading equipped items that have attachments
+				///
+				/// The infinite item upgrade is the only firing that does not contain an EventData
+				`XEVENTMGR.TriggerEvent('ItemUpgraded', UpgradedItemState, none, NewGameState);
```
Comment (Iridar) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc:4664]: This should probably have event docs, for example:
```
/// ```event
/// EventID: ItemUpgraded,
/// EventData: XComGameState_Item,
/// EventSource: none,
/// NewGameState: yes
/// ```
```
Also, when you have "```event", the "events" tag is not necessary.

Although not sure how event docs should look like, seeing as you use the same EventID, but with different EventData / EventSource availability.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc [hunk 5]
```diff
@@ -4668,6 +4679,7 @@ static function UpgradeItems(XComGameState NewGameState, name CreatorTemplateNam
 					// Otherwise match the base items quantity
 					UpgradedItemState = UpgradeItemTemplate.CreateInstanceFromTemplate(NewGameState);
 					UpgradedItemState.Quantity = BaseItemState.Quantity;
+					`XEVENTMGR.TriggerEvent('ItemUpgraded', UpgradedItemState, BaseItemState, NewGameState); // Issue #289 - handle upgrades for utility items like grenades and medkits
```
Comment (Iridar) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_HeadquartersXCom.uc:4682]: Probably should mark other trigger events with single line HLDocs:

/// HL-Docs: ref:ItemUpgraded

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc [hunk 6]
```diff
@@ -2363,9 +2363,42 @@ function bool IsMissionObjectiveItem()
 	return false;
 }
 
+// Start Issue #93
+/// HL-Docs: feature:OverrideNumUpgradeSlots; issue:93; tags:strategy,events
+/// This is an event that mods can listen for which will allow them to
+/// modify upgrade slots on a per-gamestate basis.
+/// 
+/// In general, it should only be used if you have some type of handling that allows
```
Comment (Iridar) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc:2371]: Please add event docs to this, for example:
```
/// ```unrealscript
/// EventID: OverrideLostSpawnIncreaseFromUse
/// EventData: XComLWTuple {
/// 	Data: [
///       inout int LostSpawnModifier,
///       in XComGameState_Ability ActivatedAbilityState,
///     ]
/// }
///	EventSource: self (XComGameState_Unit)
/// GameState: no
/// ```
```

File: docs_src/docs/dlcinfo.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<h1>X2DownloadableContentInfo Hooks</h1>
+
+Many of the Highlander's features utilize function calls to the subclasses of
```
Comment (Iridar) [inline docs_src/docs/dlcinfo.md:3]: Not sure this is supposed to be here.

**Review by Iridar — CHANGES_REQUESTED** [review]
Had to locally rebase this onto the current master to build it.

I have tested the features in this PR and they seem to be working fine, although upgrading non-weapon items requires an external UI mod, such as Mod Everything Reloaded.

I did have a few nitpicks to documentation and implementation, but other than that this is good to go.

**Comment by Iridar** [comment]
Closing this PR in favor of #941, which ships the same changes with better docs.

**Pull request closed without merging**